app/services/orphan_s3_video_derivatives.rb

```diff
@@ -0,0 +1,178 @@
+
+  attr_reader :s3_iterator, :shrine_storage, :show_progress_bar, :sample, :orphans_found
+  def initialize(show_progress_bar: true)
+
+    @sample = []
+    @shrine_storage = ScihistDigicoll::Env.shrine_video_derivatives_storage
+    @show_progress_bar = show_progress_bar
+
+    @s3_iterator = S3PathIterator.new(
+      extra_prefix: 'hls/',
+      first_level_only: true,
+      shrine_storage: shrine_storage,
+      show_progress_bar: show_progress_bar,
+      progress_bar_total: video_derivative_count,
+    )
+  end
+
+  def video_derivative_count
+    @video_derivative_count ||= Kithe::Asset.connection.select_one("select count((json_attributes->'hls_playlist_file_data' #>> '{}')::jsonb->>'id') from kithe_models where (json_attributes->'hls_playlist_file_data' #>> '{}')::jsonb->>'id' !='';")['count']
+  end
+
+  # Lists any orphan HLS derivatives found.
+  def report_orphans
+    max_reports = 20
+    @orphans_found = 0
+    files_checked = find_orphan_video_derivatives do |asset_id:, shrine_id:, s3_path:|
+      @orphans_found +=1
+      if @orphans_found == max_reports
+        s3_iterator.log "Reported max #{max_reports} orphans, not listing subsquent...\n"
+      elsif @orphans_found < max_reports
+        @sample << s3_url_for_path(s3_path)
+        asset = Asset.where(id: asset_id).first
+        s3_iterator.log "set of orphaned HLS derivatives"
+        s3_iterator.log "  bucket: #{shrine_storage.bucket.name}"
+        s3_iterator.log "  s3 path: #{s3_path}"
+        s3_iterator.log "  asset_id: #{asset_id}"
+        s3_iterator.log "  asset missing"  if asset.nil?
+        s3_iterator.log ""
+      end
+    end
+
+    $stderr.puts "\n\nTotal video derivative count from the database: #{video_derivative_count}"
+    $stderr.puts "Iterated through #{files_checked} video derivatives on S3"
+    $stderr.puts "Found #{orphans_found} orphaned items\n"
+  end
+
+  # Delete all files and folders that can't be accounted for
+  # as HLS derivatives.
+  def delete_orphans
+    delete_count = 0
+    find_orphan_video_derivatives do |asset_id:, shrine_id:, s3_path:|
+      shrine_storage.delete_prefixed(shrine_id)
+      s3_iterator.log "deleted: #{shrine_storage.bucket.name}: #{s3_path}"
+      delete_count += 1
+    end
+    $stderr.puts "\nDeleted #{delete_count} sets of orphaned derivatives"
+  end
+
+
+  # Yields to the caller an s3 path for each hls derivative folder on s3.
+  # We expect each of these to contain exactly one `hls.m3u8` manifest file,
+  # as well as dozens to hundreds of actual derivative files.
+  def all_hls_deriv_folders
+    # First iterate over all video derivs in s3.
+    s3_iterator.each_s3_path do |all_derivs_for_asset|
+      # all_derivs_for_asset is an s3 folder containing
+      # all HLS derivatives for a particular asset.
+      # The path ends in the asset id, e.g.
+      # laptop.local/derivatives_video/hls/0eadfd9b-c36b-4b02-a5e1-63a00a259410/
+      if /\/$/.match? all_derivs_for_asset
```

Comment:
I don't understand what this condition is matching. The `s3_iterator` above is iterating over S3 and yielding... `all_derivatives_for_asset`?  I don't understand what this means in the first place.... it is yielding one path at a time, because that's what an s3_iterator does, right?  How can a single path be `all_derivs_for_asset`, what does that variable name mean?  

And then you are testing whether `all_derivs_for_asset` ends in a `/`? I don't understand why, and also if I'm right about what it's doing that can be much more clearly written as `all_derivs_for_asset.end_with?('/')` I think?

Reply:
all_derivs_for_asset.end_with?('/') is much better yeah. And fixed the variable name.